chalk-integration/src/program_writer.rs

```diff
@@ -0,0 +1,980 @@
+use chalk_solve::display::{RenderAsRust, WriterState};
+
+use crate::program::Program;
+
+pub trait WriteProgram {
+    fn write(&self) -> String;
+}
+
+impl WriteProgram for Program {
+    fn write(&self) -> String {
+        let mut lines = vec![];
+        let ws = &WriterState::new(self);
+        self.struct_data.values().for_each(|datum| {
+            lines.push(datum.display(ws).to_string());
+        });
+        self.trait_data.values().for_each(|datum| {
+            lines.push(datum.display(ws).to_string());
+        });
+        self.impl_data.values().for_each(|datum| {
+            lines.push(datum.display(ws).to_string());
+        });
+        lines.join("\n")
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
These should probably live in the [`tests`](https://github.com/rust-lang/chalk/tree/master/tests) directory

Reply:
Thanks! Right now this file is mostly just a stopgap so we could write tests before figuring out the tests directory structure.